### PR TITLE
Allow workers to shut down cleanly after processing in-progress messages

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

File: src/vumi2/connectors.py [hunk 1]
```diff
@@ -73,28 +77,49 @@ async def start(self) -> None:
...
-            except Exception as e:
-                logger.exception(e)
-                await channel.basic_client_nack(
-                    delivery_tag=envelope.delivery_tag, requeue=False
-                )
-
-
-class Publisher:
+        if not self._closing:
+            await self.send_channel.send((channel, body, envelope, properties))
```
Comment (rudigiesler) [inline src/vumi2/connectors.py:98]: Should we raise something if we are closing? I'm worried that something gets marked as completed successfully here, but nothing was actually queued.

**Review by rudigiesler — COMMENTED** [review]
Just one question, otherwise looks good!

File: src/vumi2/connectors.py (hunk 1 above)
Comment (jerith) [reply src/vumi2/connectors.py:98]: If we never send the message to a handler, it never gets acked/nacked and should thus be requeued by the broker when we close the AMQP channel. (Unless we've told the broker to treat the message as delivered as soon as it reaches a consumer, in which case it's already too late to do anything here anyway.)

**Comment by jerith** [comment]
Something I just thought of: We should probably finish closing all of a worker's consumers before we start closing any of its publishers, otherwise we may no be able to send replies to in-progress messages. That means `AsyncResource` may not be an entirely appropriate API for connectors. 🤔

File: src/vumi2/connectors.py (hunk 1 above)
Comment (jerith) [reply src/vumi2/connectors.py:98]: We really should have tests for this sort of thing, though. I'll see about writing some.

**Comment by jerith** [comment]
I'm not sure this test work really belongs in this PR, but it's not really practical to extract it. 😕

**Comment by jerith** [comment]
There may still be some tests that are worth writing, but we can do those later when we think of them.

**Review by rudigiesler — APPROVED** [review]
Looks good! I can't think of any tests that might be missing.

The only bit that was a bit confusing at first was the `AcloseWorker` and all its channels. I wonder if a comment there describing the flow of messages through the channels, and how we use that to test shutdown, might help?

**Review by rudigiesler — APPROVED** [review]

**Pull request merged**